### ISSUE TITLE
Show error message if library is missing

### ISSUE THREAD

**Issue by probonopd** [mate-desktop/caja#1262]
#### Expected behaviour

An error box is shown, like in Haiku.

![](https://user-images.githubusercontent.com/2480569/61177311-aa8ab900-a5c1-11e9-8130-804d01ec37a2.png)

#### Actual behaviour

Nothing happens

#### Steps to reproduce the behaviour

Double-click an ELF binary that has dependencies which are missing on the system

#### MATE general version

Any

#### Package version

Any

#### Linux Distribution

Any

#### Link to downstream report of your Distribution

None

#### Cross-references

* GNOME: https://gitlab.gnome.org/GNOME/nautilus/issues/1166
* KDE: https://bugs.kde.org/show_bug.cgi?id=409777
* XFCE: https://bugzilla.xfce.org/show_bug.cgi?id=15714

**Comment by Max-E** [comment]
I might give this a shot if no one else gets to it before the weekend.

It looks like the right place for this functionality to be added is `caja_launch_application_from_command` in caja-program-choosing.c, if `use_terminal == false`.

If, for whatever reason, you didn't want this to apply to things like custom right-click scripts, launchers in the sidepane, internal MATE executables, etc., you could put it in `activate_files` in caja-mime-actions.c.

**Comment by probonopd** [comment]
@Max-E thanks for having a go at this. Let me know if I can test something.

**Comment by Max-E** [comment]
Did you have an opinion about whether this should be narrowly limited to just double-clicking executables in the file view, or other executables that get invoked by Caja as well? I did read your Medium post and you've clearly spent a lot of time thinking about UX. As far as I can tell, the pros and cons of applying it _everywhere_ are

Pros:
* User is alerted to internal problems with MATE's components and is given the opportunity to fix them
* Things like custom caja addons are now easier to troubleshoot
Cons:
* Wider splash damage = higher regression risk (i.e. the conservatism argument)
* A problem that was previously benign and asymptomatic now turns into a Hell of incessant pop-up dialogs

**Comment by flexiondotorg** [comment]
Why should the file manager be used to confirm the integrity of installed software? Surely this is the responsibility of the operating systems package manager?

**Comment by Max-E** [comment]
For easy testing purposes, I have written a short shell script that will generate a Linux executable with missing dependencies. This is a quick and repeatable way to test this feature without trying to copy over an executable from another system or something.  
https://gist.github.com/Max-E/4a9c4c3f982d5ab952c0d89e9def1066

**Comment by probonopd** [comment]
>  narrowly limited to just double-clicking executables in the file view

Yes, I think this would do. This feature is for helping users to sideload software, not for debugging the OS itself.

>  Why should the file manager be used to confirm the integrity of installed software? Surely this is the responsibility of the operating systems package manager?

This is especially useful for software that was sideloaded without using the system's package manager.

**Comment by rbuj** [comment]
https://en.wikipedia.org/wiki/Rpath

**Comment by Max-E** [comment]
Annoyingly, due to this weirdness: 
https://stackoverflow.com/questions/41398444/gcc-creates-mime-type-application-x-sharedlib-instead-of-application-x-applicati 
Some executables get detected as "application/x-sharedlib" which prevents them from running when double-clicked, even if you DO have all the dependencies. For instance, when I double-click /bin/ls in the file manager (an executable that assuredly works just fine,) I get a pop-up saying "There's no application installed for shared library files."

I'm going to treat that as out-of-scope for the purposes of this ticket, and will only attempt to fix the behavior for non-position-independent executables that currently _can_ be launched through caja, if the dependencies for _those_ are missing.

**Comment by Max-E** [comment]
Looks like while there are plenty of command line programs to detect and list missing dependencies, there's nothing like a C library to do it. For instance, `ldd` seems to work by calling `/lib64/ld-linux-x86-64.so.2`, which is a statically linked executable that cannot have its code reused via `dlopen`. So my options are either write something myself, or shell out to an existing program.

**Comment by lukefromdc** [comment]
BTW, non-PIE files are more vulnerable to any buffer overflow attacks, keep that in mind for any machine with potential multiple users, including websites running javascript

**Comment by rbuj** [comment]
application/x-pie-executable is displayed as "shared library" (https://cgit.freedesktop.org/xdg/shared-mime-info/tree/freedesktop.org.xml.in#n3552) in Type column instead of "executable" (https://cgit.freedesktop.org/xdg/shared-mime-info/tree/freedesktop.org.xml.in#n1982)
```shell
$ file --mime-type -b /bin/ls
application/x-pie-executable
$ file --mime-type -b /bin/lsgpio 
application/x-executable
```

![Screenshot at 2019-08-12 10-51-58](https://user-images.githubusercontent.com/10171411/62854332-57567400-bcef-11e9-9448-e483f7b33ca4.png)

**Comment by Max-E** [comment]
I can confirm that `file` behaves as you say, but the `mimetype` command does show "application/x-sharedlib" on my system, and caja does show the pop-up instead of actually running `ls`. I would believe it's a packaging problem at the distro level, which is why I'm not pursuing it further right now.

**Comment by probonopd** [comment]
Could the error the loader(?) dumps to the command line when a library is missing be caught and shown in the GUI? The way it is now feels like an "unhandled exception" to me, an error that is dumped to stderr which the user of a GUI app will most likely never have a chance to see. Can we "handle the exception"?

**Comment by probonopd** [comment]
> there's nothing like a C library to do it 

How is Haiku doing it? They, too, are using the ELF format for binaries, so one way or the other they must already have solved this.

Here is their C++ code in question:

https://github.com/haiku/haiku/blob/f7397837e8e8749d0a2cdf2725262f00ccd74995/src/kits/tracker/FSUtils.cpp#L3658-L3693

So the file manager in Haiku reacts to `B_MISSING_LIBRARY` and `B_MISSING_SYMBOL` errors.

The `B_MISSING_LIBRARY` error gets set here:
https://github.com/haiku/haiku/blob/da73ed96dd833549c0d24544a5113e942f7e6be6/src/system/runtime_loader/elf.cpp#L195

The `B_MISSING_SYMBOL` error gets set here:
https://github.com/haiku/haiku/blob/a9f4ce4515dfe036de4eca055ca8f83da4548abc/src/system/boot/loader/elf.cpp#L445

Possibly @weinhold could give us some additional hints on how to implement this best on Linux.

**Comment by Max-E** [comment]
Detecting it after the fact is kind of hard without some pretty fragile code. It's hard to distinguish from other exiting programs reliably. There's no special error code for it. If you try to recognize it by text, what happens if you run it on a non-English system? It's also hard to pick a reasonable cutoff time for the detection that makes sense on both a HDD and an SSD.

I think the right place for this might be glibc honestly. That's where the ld.so sources are kept, so that code could be repurposed fairly easily for the purpose, instead of writing yet another new ELF parser. I recently discovered that ld.so is secretly the executable containing ldd's functionality, so it's basically already there. Just not in a form you can link to and use.

**Comment by probonopd** [comment]
> I think the right place for this might be glibc honestly. 

:+1: 

Now, that would be one hell of a feature for glibc. Unfortunately the glibc project is not on GitHub so we can't just open an issue there. Anyone knows how to suggest it to that project "properly"?

**Comment by xordspar0** [comment]
Did anyone ever report this issue in glibc? If so I can't find it. I would do it, but I'm not sure what to ask for. What do we need from ld.so? Today it prints a message to stderr and exits with code 127.

**Comment by probonopd** [comment]
@xordspar0 I didn't because glibc is not on GitHub Issues... one needs to sign up for https://sourceware.org/bugzilla/

**Comment by Max-E** [comment]
I envisioned glibc adding a header with functions to query the same information from an executable that ldd can query from the command line. This allows us to avoid error-prone C code to execute and parse the output of the ldd command line tool.
